src/platform/OpenThread/GenericThreadBorderRouterDelegate.cpp

```diff
@@ -0,0 +1,197 @@
+    if (type == DatasetType::kActive)
+    {
+        otErr = otDatasetGetActiveTlvs(otInst, &datasetTlvs);
+    }
+    else
+    {
+        otErr = otDatasetGetPendingTlvs(otInst, &datasetTlvs);
+    }
+    if (otErr == OT_ERROR_NONE)
+    {
+        return dataset.Init(ByteSpan(datasetTlvs.mTlvs, datasetTlvs.mLength));
+    }
+    return CHIP_ERROR_NOT_FOUND;
+}
+
+CHIP_ERROR GenericOpenThreadBorderRouterDelegate::SetActiveDataset(const Thread::OperationalDataset & activeDataset,
+                                                                   uint32_t sequenceNum, ActivateDatasetCallback * callback)
+{
+    // This function will never be invoked when there is an Active Dataset already configured.
+    CHIP_ERROR err = SaveThreadBorderRouterCommissioned(false);
+    if (err == CHIP_NO_ERROR)
+    {
+        err = DeviceLayer::ThreadStackMgrImpl().AttachToThreadNetwork(activeDataset, nullptr);
+    }
+    if (err == CHIP_NO_ERROR)
+    {
+        // We can change the stored sequence number because OnActivateDatasetComplete executed later must be called for this
+        // SetActiveDataset request
+        mSequenceNum = sequenceNum;
+        mCallback    = callback;
+    }
+    return err;
+}
+
+void GenericOpenThreadBorderRouterDelegate::OnPlatformEventHandler(const DeviceLayer::ChipDeviceEvent * event, intptr_t arg)
+{
+    GenericOpenThreadBorderRouterDelegate * delegate = reinterpret_cast<GenericOpenThreadBorderRouterDelegate *>(arg);
+    if (delegate && delegate->mCallback)
+    {
+        if (event->Type == DeviceLayer::DeviceEventType::kThreadConnectivityChange &&
+            event->ThreadConnectivityChange.Result == DeviceLayer::kConnectivity_Established)
+        {
+            delegate->mCallback->OnActivateDatasetComplete(delegate->mSequenceNum, CHIP_NO_ERROR);
+            delegate->mCallback = nullptr;
+        }
+    }
+}
+
+CHIP_ERROR GenericOpenThreadBorderRouterDelegate::SaveThreadBorderRouterCommissioned(bool commissioned)
+{
+    return DeviceLayer::PersistedStorage::KeyValueStoreMgr().Put(kFailsafeThreadBorderRouterCommissioned, commissioned);
+}
+
+CHIP_ERROR GenericOpenThreadBorderRouterDelegate::CommitActiveDataset()
+{
+    return SaveThreadBorderRouterCommissioned(true);
+}
+
+CHIP_ERROR GenericOpenThreadBorderRouterDelegate::RevertActiveDataset()
+{
+    // The FailSafe Timer is triggered and the previous command request should be handled, so reset the callback.
+    mCallback               = nullptr;
+    bool threadCommissioned = true;
+    DeviceLayer::PersistedStorage::KeyValueStoreMgr().Get(kFailsafeThreadBorderRouterCommissioned, &threadCommissioned);
+    if (!threadCommissioned)
+    {
+        // If Thread is not commissioned, we will try to attach an empty Thread dataset and the Thread Border Router
+        // will stay uncommissioned
+        Thread::OperationalDataset emptyDataset = {};
+        return DeviceLayer::ThreadStackMgrImpl().AttachToThreadNetwork(emptyDataset, nullptr);
```

Comment:
If I understand correctly, the intent here is to clear the active dataset saved in the OT stack since the Fail Safe timer was triggered. Will this work ? While looking at the dataset API it seems there is no API to clear an Active Dataset.  

Maybe we should update a bit the comment to explain what we are actually trying to do:`If Thread is not commissioned, we will try to attach using an empty Thread dataset and that will clear the one stored in the Thread stack since the SetActiveDataset operation failed`.

Reply:
Done